fw/http.c

```diff
@@ -3412,214 +3374,179 @@ tfw_h1_set_loc_hdrs(TfwHttpMsg *hm, bool is_resp, bool from_cache)
 		 * header-by-header right away.
 		 */
 		if (from_cache) {
-			TfwHttpResp *resp = (TfwHttpResp *)hm;
-			struct sk_buff **skb_head = &resp->msg.skb_head;
-			TfwHttpTransIter *mit = &resp->mit;
-			TfwStr crlf = { .data = S_CRLF, .len = SLEN(S_CRLF) };
+			struct sk_buff **skb_head = &hm->msg.skb_head;
+			TfwMsgIter *it = &hm->iter;
 			/*
 			 * Skip the configured header if the header is
 			 * configured for deletion (without value chunk).
 			 */
 			if (h_mdf.nchunks < 3)
 				continue;
 			/* h_mdf->eolen is ignored, add explicit CRLF. */
-			r = tfw_http_msg_expand_data(&mit->iter, skb_head,
-						     &h_mdf, NULL);
-			r |= tfw_http_msg_expand_data(&mit->iter, skb_head,
-						     &crlf, NULL);
+			r = tfw_http_msg_expand_data(it, skb_head, &h_mdf,
+						     NULL);
+			if (unlikely(r))
+				return r;
+			r |= tfw_http_msg_expand_data(it, skb_head, &crlf,
+						      NULL);
 		} else {
-			r = tfw_http_msg_hdr_xfrm_str(hm, &h_mdf, d->hid,
-						      d->append);
-		}
-
-		if (r) {
-			T_ERR("can't update location-specific header in msg %p\n",
-			      hm);
-			return r;
+			r = tfw_http_msg_expand_from_pool(hm, &h_mdf);
+			if (unlikely(r))
+				return r;
+			r = tfw_http_msg_expand_from_pool(hm, &crlf);
 		}
-
-		T_DBG2("updated location-specific header in msg %p\n", hm);
 	}
 
-	return 0;
+	return r;
 }
 
+/**
+ * Adjust the request before proxying it to real server.
+ */
 static int
-tfw_h1_add_loc_hdrs(TfwHttpMsg *hm, const TfwHdrMods *h_mods)
+tfw_h1_adjust_req(TfwHttpReq *req)
 {
-	unsigned int i;
-	TfwHttpHdrTbl *ht = hm->h_tbl;
+	int r, cl_sz = sizeof(TfwHttpMsgCleanup);
+	TfwHttpMsg *hm = (TfwHttpMsg *)req;
+	TfwHttpMsgCleanup *cleanup = req->cleanup;
+	const BasicStr *s_meth;
+	TfwStr *pos, *end, meth = {};
+	static const DEFINE_TFW_STR(sp, " ");
+	static const DEFINE_TFW_STR(crlf, S_CRLF);
+	static const DEFINE_TFW_STR(ver, " " S_VERSION11 S_CRLF);
+	const TfwHdrMods *h_mods = tfw_vhost_get_hdr_mods(req->location,
+							  req->vhost,
+							  TFW_VHOST_HDRMOD_REQ);
 
-	if (!h_mods)
-		return 0;
+	cleanup = (TfwHttpMsgCleanup *)tfw_pool_alloc(hm->pool, cl_sz);
+	memset(cleanup, 0, cl_sz);
 
-	for (i = 0; i < h_mods->sz; ++i) {
-		const TfwHdrModsDesc *desc = &h_mods->hdrs[i];
-		int r;
-		unsigned short hid = desc->hid;
-		TfwStr h_mdf = {
-			.chunks = (TfwStr []){
-				{},
-				{ .data = S_DLM, .len = SLEN(S_DLM) },
-				{}
-			},
-			.len = SLEN(S_DLM),
-			.nchunks = 2 /* header name + delimeter. */
-		};
+	tfw_msg_transform_setup(&req->iter, req->msg.skb_head);
+	r = tfw_http_msg_cutoff_headers(hm, cleanup);
+	if (unlikely(r))
+		goto clean;
 
-		/* FIXME: this is a temporary WA for GCC12, see #1695 for details */
-		if (TFW_STR_CHUNK(desc->hdr, 1) == NULL)
-			continue;
+	if (test_bit(TFW_HTTP_B_PURGE_GET, req->flags))
+		/* Rewrite PURGE to GET */
+		s_meth = tfw_http_method_id2str(TFW_HTTP_METH_GET);
+	else
+		s_meth = tfw_http_method_id2str(req->method);
```

Comment:
Can req->method be greater then size of static array in " tfw_http_method_id2str"?

Reply:
It's a good point, I checked we allow unknown methods and it implies I can't use static array I must store method during parsing.